### DISCUSSION TITLE
Active/active multisite for sharedfs, is it possible

### DISCUSSION THREAD

**Discussion by ecooke-macu** [rook/rook#14890] [Q&A]
We're rebuilding our clusters and I'm wondering if we can use Ceph to solve our storage needs. We will have 2 clusters, in 2 separate geographic locations (~7 ms latency). We have many applications that run in both clusters that need access to shared storage. Currently we achieve this with NFS in Azure and the nfs provisioner. Less than ideal. We want to run it in local clusters.

Is there a way of doing this same thing with Rook? Looking at it, it seams that Ceph can do it with multisite, there's a blurb in the docs that state active/active is possible with writes to both replicas, but there's no docs on how to achieve that. I also can't seem to find anything in the examples or docs of rook stating how to do an active/active replica. I'm also struggling to find how to reference the storage that is replicated from site A in site B.

**Comment by BlaineEXE** [comment]
If you are talking about NFS, I assume the use case is for shared filesystem (not block or S3 storage). In that case, filesystem mirroring may be what you are looking for. Rooks doc is here: https://rook.io/docs/rook/latest-release/Storage-Configuration/Shared-Filesystem-CephFS/filesystem-mirroring/?h=mirror

I would also suggest checking out Ceph's docs to understand more about Ceph's features/limitations when considering your needs: https://docs.ceph.com/en/reef/dev/cephfs-mirroring/

**Comment by travisn** [comment]
For active/active, consider using a single cluster, then have a CephFS RXW volume that is shared between the two active applications. The latency is low enough that a single K8s cluster should not be a concern. 

The challenge with replication across clusters is it's async so it's more of an active/passive scenario.

**Comment by jhoblitt** [comment]
In theory, it should be possible to configure multi-site rgw and then export that as nfs. Note that when exporting rgw via nfs that it does not conform to the standard expectations of a posix filesystem. See https://docs.ceph.com/en/latest/radosgw/nfs/#supported-operations

AFS (Andrew Filesystem) has support for active/active replication but also deviates in its behavior from a standard posix filesystem.

My suggestion would be to consider how the application will address CAP theorem limitations and to select a state management strategy that is able to to fulfill those requirements.

**Reply by ecooke-macu** [reply]
Is it the RGW that doesn't support the random writes? Or is it the NFS server that doesn't? I could expose the RGW using another NFS server other than the one included within Ceph/Rook, maybe. But if it's the RGW that is the limiting factor that obviously wouldn't work.

For active/active this is the reference to it I found in Ceph documentation.

https://docs.ceph.com/en/reef/radosgw/multisite/#functional-changes-from-infernalis

**Reply by jhoblitt** [reply]
The AWS s3 API does not provide a mechanism for partial updates to objects. RGW provides the AWS s3 API and has the same limitation.

**Reply by ecooke-macu** [reply]
Thanks for that information.

**Comment by ecooke-macu** [comment]
We have to use 2 separate kubernetes clusters due to DR constraints, we have to be fully operational in the event that a datacenter goes down. We also do tests annually to make sure that is working as expected. We could do some trickery in Kubernetes to do that with a single cluster but it becomes a headache that we opted to not deal with when it comes to the control plane/etcd.

As for exposing it over NFS that could be a problem as I believe we do random writes/updates to the files where that looks like a limitation in NFS based on the link supplied by @jhoblitt. So that's sad, that was what I was thinking of doing. Having everything write to one cluster, and during a failover, it would automatically switch to the second. Which is a viable option as long as it's automatic. I'll take a look at AFS. Is that supported by Rook? Or is that something different?

**Reply by jhoblitt** [reply]
AFS is not supported by rook.

**Reply by ecooke-macu** [reply]
Thanks.

**Comment by ecooke-macu** [comment]
So it seems like what I may end up doing is this
object store block device running in k8s cluster 1 mirrored to k8s cluster 2
some nfs server running on k8s cluster 1 with dns entry (nfs.k8s)
cluster 1/cluster 2 use nfs mount on nfs.k8s
failover would involve manual steps to bring k8s cluster2 object store active and nfs server running on k8s cluster 2 and switch dns entry to k8s cluster 2. Restart all pods that use storage as well.

Since it sounds like I'm going to need to do active/passive (sad) does the above sound about right in a way to use rook/ceph?

**Reply by BlaineEXE** [reply]
Failover (and failback) are incredibly challenging things to orchestrate within the NFS protocol. Restarting all storage-consuming pods could help with that though.

Active/active NFS between 2 sites will likely never be supported by Ceph. I'm aware that the Ceph team is working on some features that may allow active/active for a single site, but those have been delayed several times, so I don't have a roadmap for them.

As Travis mentioned, stretching a single K8s cluster across both sites may end up being easier.
